File: languages/python/oso/polar/host.py
"""Translate between Polar and the host language (Python)."""

from dataclasses import dataclass
from math import inf, isnan, nan
import re
import inspect
from typing import Any, Dict, Union


from .exceptions import (
    PolarRuntimeError,
    UnregisteredClassError,
    DuplicateClassAliasError,
    UnregisteredInstanceError,
    DuplicateInstanceRegistrationError,
    UnexpectedPolarTypeError,
    UNEXPECTED_EXPRESSION_MESSAGE,
)
from .variable import Variable
from .predicate import Predicate
from .expression import Expression, Pattern
from .data_filtering import Relation


@dataclass
class UserType:
    name: str
    cls: type
    id: int
    fields: Dict[str, Any]


class Host:
    """Maintain mappings and caches for Python classes & instances."""

    types: Dict[Union[str, type], UserType]

    def __init__(
        self,
        polar,
        types=None,
        instances=None,
        get_field=None,
        adapter=None,
    ):
        assert polar, "no Polar handle"
        self.ffi_polar = polar  # a "weak" handle, which we do not free
        # types maps class names (as string) and class objects to UserType.
        self.types = (types or {}).copy()
        self.instances = (instances or {}).copy()
        self._accept_expression = False  # default, see set_accept_expression
        self.adapter = adapter

        self.get_field = get_field or self.types_get_field

    # @Q: I'm not really sure what I'm returning here.
    def types_get_field(self, obj, field) -> type:
        if obj not in self.types:
            raise PolarRuntimeError(
                f"No type information for Python class {obj.__name__}"
            )
        rec = self.types[obj]

        if field not in rec.fields:
            raise PolarRuntimeError(f"No field {field} on {obj.__name__}")
        field_type = rec.fields[field]

        if not isinstance(field_type, Relation):
            return field_type

        if field_type.kind == "one":
            return self.types[field_type.other_type].cls
        elif field_type.kind == "many":
            return list
        else:
            raise PolarRuntimeError(f"Invalid kind {field_type.kind}")

    def copy(self) -> "Host":
        """Copy an existing cache."""
        return type(self)(
            self.ffi_polar,
            types=self.types,
            instances=self.instances,
            get_field=self.get_field,
            adapter=self.adapter,
        )

    def get_class(self, name):
        """Fetch a Python class from the cache."""
        try:
            return self.types[name].cls
        except KeyError:
            raise UnregisteredClassError(name)

    def distinct_user_types(self):
        return map(
            lambda k: self.types[k],
            filter(lambda k: isinstance(k, str), self.types.keys()),
        )

    def cache_class(
        self,
        cls,
        name=None,
        fields=None,
    ):
        """Cache Python class by name."""
        name = cls.__name__ if name is None else name
        if name in self.types.keys():
            raise DuplicateClassAliasError(name, self.get_class(name), cls)

        self.types[name] = self.types[cls] = UserType(
            name=name,
            cls=cls,
            id=self.cache_instance(cls),
            fields=fields or {},
        )
        return name

    def register_mros(self) -> None:
        """Register the MRO of each registered class to be used for rule type validation."""
        # Get MRO of all registered classes
        for rec in self.distinct_user_types():
            mro = [self.types[c].id for c in inspect.getmro(rec.cls) if c in self.types]
            self.ffi_polar.register_mro(rec.name, mro)

    def get_instance(self, id):
        """Look up Python instance by id."""
        if id not in self.instances:
            raise UnregisteredInstanceError(id)
        return self.instances[id]

    def cache_instance(self, instance, id=None):
        """Cache Python instance under Polar-generated id."""
        if id is None:
            id = self.ffi_polar.new_id()
        self.instances[id] = instance
        return id

    def make_instance(self, name, args, kwargs, id):
        """Construct and cache a Python instance."""
        if id in self.instances:
            raise DuplicateInstanceRegistrationError(id)
        cls = self.get_class(name)
        try:
            instance = cls(*args, **kwargs)
        except Exception as e:
            raise PolarRuntimeError(f"Error constructing instance of {name}: {e}")
        return self.cache_instance(instance, id)

    def unify(self, left_instance_id, right_instance_id) -> bool:
        """Return true if the left instance is equal to the right."""
        left = self.get_instance(left_instance_id)
        right = self.get_instance(right_instance_id)
        return left == right

    def isa(self, instance, class_tag) -> bool:
        instance = self.to_python(instance)
        cls = self.get_class(class_tag)
        return isinstance(instance, cls)

    def isa_with_path(self, base_tag, path, class_tag) -> bool:
        base = self.get_class(base_tag)
        cls = self.get_class(class_tag)
        for field in path:
            field = self.to_python(field)
            base = self.get_field(base, field)
        return issubclass(base, cls)

    def is_subclass(self, left_tag, right_tag) -> bool:
        """Return true if left is a subclass (or the same class) as right."""
        left = self.get_class(left_tag)
        right = self.get_class(right_tag)
        return issubclass(left, right)

    def is_subspecializer(self, instance_id, left_tag, right_tag) -> bool:
        """Return true if the left class is more specific than the right class
        with respect to the given instance."""
        try:
            mro = self.get_instance(instance_id).__class__.__mro__
            left = self.get_class(left_tag)
            right = self.get_class(right_tag)
            return mro.index(left) < mro.index(right)
        except ValueError:
            return False

    def operator(self, op, args):
        try:
            if op == "Lt":
                return args[0] < args[1]
            elif op == "Gt":
                return args[0] > args[1]
            elif op == "Eq":
                return args[0] == args[1]
            elif op == "Leq":
                return args[0] <= args[1]
            elif op == "Geq":
                return args[0] >= args[1]
            elif op == "Neq":
                return args[0] != args[1]
            else:
                raise PolarRuntimeError(
                    f"Unsupported external operation '{type(args[0])} {op} {type(args[1])}'"
                )
        except TypeError:
            raise PolarRuntimeError(
                f"External operation '{type(args[0])} {op} {type(args[1])}' failed."
            )

    def enrich_message(self, message: str) -> str:
        """
        "Enrich" a message from the polar core, such as a log line, debug
        message, or error trace.

        Currently only used to enrich messages with instance reprs. This allows
        us to avoid sending reprs eagerly when an instance is created in polar.
        """

        def replace_repr(match):
            instance_id = int(match[1])
            try:
                instance = self.get_instance(instance_id)
                return repr(instance)
            except UnregisteredInstanceError:
                return match[0]

        return re.sub(r"\^\{id: ([0-9]+)\}", replace_repr, message, flags=re.M)

    def to_polar(self, v):
        """Convert a Python object to a Polar term."""
        if type(v) == bool:
            val = {"Boolean": v}
        elif type(v) == int:
            val = {"Number": {"Integer": v}}
        elif type(v) == float:
            if v == inf:
                v = "Infinity"
            elif v == -inf:
                v = "-Infinity"
            elif isnan(v):
                v = "NaN"
            val = {"Number": {"Float": v}}
        elif type(v) == str:
            val = {"String": v}
        elif type(v) == list:
            val = {"List": [self.to_polar(i) for i in v]}
        elif type(v) == dict:
            val = {
                "Dictionary": {"fields": {k: self.to_polar(v) for k, v in v.items()}}
            }
        # only used when you call oso.query() with a Predicate instance
        elif isinstance(v, Predicate):
            val = {
                "Call": {
                    "name": v.name,
                    "args": [self.to_polar(v) for v in v.args],
                }
            }
        # basically only used in data filtering or if someone intentionally manually passes in a Variable instance
        elif isinstance(v, Variable):
            val = {"Variable": v}
        # basically only used in data filtering
        elif isinstance(v, Expression):
            val = {
                "Expression": {
                    "operator": v.operator,
                    "args": [self.to_polar(v) for v in v.args],
                }
            }
        # basically only used in data filtering (seeding the authorized_query()
        # call with an initial type binding so we know what type of resources
        # we're trying to determine access for)
        elif isinstance(v, Pattern):
            if v.tag is None:
                val = {"Pattern": self.to_polar(v.fields)["value"]}
            else:
                val = {
                    "Pattern": {
                        "Instance": {
                            "tag": v.tag,
                            "fields": self.to_polar(v.fields)["value"]["Dictionary"],
                        }
                    }
                }

        # user queries: oso.allow(<some user instance>, "some action", <some resource instance>)
        # Host.to_polar translates that into something like
        #   Call {
        #       name: String("allow"),
        #       args: List([
        #           ExternalInstance { instance_id: 1, repr: "<some user instance>", class_id: <user_class_id> },
        #           String("some action"),
        #           ExternalInstance { instance_id: 2, repr: "<some resource instance>", class_id: <resource_class_id> },
        #       ]
        #   }
        else:
            import inspect

            instance_id = None
            class_id = None

            # maintain consistent IDs for registered classes
            if inspect.isclass(v):
                if v in self.types:
                    class_id = instance_id = self.types[v].id

            # pass the class_repr only for registered types otherwise None
            class_repr = self.types[type(v)].name if type(v) in self.types else None

            # pass class_id for classes & instances of registered classes,
            # otherwise pass None
            if type(v) in self.types:
                class_id = self.types[type(v)].id

            val = {
                "ExternalInstance": {
                    "instance_id": self.cache_instance(v, instance_id),
                    "repr": None,
                    "class_repr": class_repr,
                    "class_id": class_id,
                }
            }
        term = {"value": val}
        return term

    def to_python(self, value):
        """Convert a Polar term to a Python object."""
        value = value["value"]
        tag = [*value][0]
        if tag in ["String", "Boolean"]:
            return value[tag]
        elif tag == "Number":
            number = [*value[tag].values()][0]
            if "Float" in value[tag]:
                if number == "Infinity":
                    return inf
                elif number == "-Infinity":
                    return -inf
                elif number == "NaN":
                    return nan
                else:
                    if not isinstance(number, float):
                        raise PolarRuntimeError(
                            f'Expected a floating point number, got "{number}"'
                        )
            return number
        elif tag == "List":
            return [self.to_python(e) for e in value[tag]]
        elif tag == "Dictionary":
            return {k: self.to_python(v) for k, v in value[tag]["fields"].items()}
        elif tag == "ExternalInstance":
            return self.get_instance(value[tag]["instance_id"])
        elif tag == "Call":
            return Predicate(
                name=value[tag]["name"],
                args=[self.to_python(v) for v in value[tag]["args"]],
            )
        elif tag == "Variable":
            return Variable(value[tag])
        elif tag == "Expression":
            if not self._accept_expression:
                raise UnexpectedPolarTypeError(UNEXPECTED_EXPRESSION_MESSAGE)

            args = list(map(self.to_python, value[tag]["args"]))
            operator = value[tag]["operator"]

            return Expression(operator, args)
        elif tag == "Pattern":
            pattern_tag = [*value[tag]][0]
            if pattern_tag == "Instance":
                instance = value[tag]["Instance"]
                return Pattern(instance["tag"], instance["fields"]["fields"])
            elif pattern_tag == "Dictionary":
                dictionary = value[tag]["Dictionary"]
                return Pattern(None, dictionary["fields"])
            else:
                raise UnexpectedPolarTypeError("Pattern: " + value[tag])

        raise UnexpectedPolarTypeError(tag)

    def set_accept_expression(self, accept):
        """Set whether the Host accepts Expression types from Polar, or raises an error."""
        self._accept_expression = accept

File: languages/ruby/lib/oso/polar/host.rb
# frozen_string_literal: true

module Oso
  module Polar
    # Ruby code reloaders (i.e. the one used by rails) swap out the value of
    # a constant on code changes. Because of this, we can't reliably call
    # `is_a?` on the constant that was passed to `register_class`.
    #
    # Example (where Foo is a class defined in foo.rb):
    #   > klass = Foo
    #   > Foo.new.is_a? klass
    #     => true
    #   > ... user changes foo.rb ...
    #   > Foo.new.is_a? klass
    #     => false
    #
    # To solve this, when we need to access the class (e.g. during isa), we
    # look it up using const_get, which will always return the up-to-date
    # version of the class.
    class PolarClass
      attr_reader :name, :anon_class

      def initialize(klass)
        @name = klass.name
        # If the class doesn't have a name, it is anonymous, meaning we should
        # actually store it directly
        @anon_class = klass if klass.name.nil?
      end

      def get
        return anon_class if anon_class

        Object.const_get(name)
      end
    end

    # For holding type metadata: name, fields, etc.
    class UserType
      attr_reader :name, :klass, :id, :fields, :build_query, :combine_query, :exec_query

      def initialize(name:, klass:, id:, fields:, build_query:, combine_query:, exec_query:) # rubocop:disable Metrics/ParameterLists
        @name = name
        @klass = klass
        @id = id
        # accept symbol keys
        @fields = fields.each_with_object({}) { |kv, o| o[kv[0].to_s] = kv[1] }
        @build_query = build_query
        @combine_query = combine_query
        @exec_query = exec_query
      end
    end

    # Translate between Polar and the host language (Ruby).
    class Host # rubocop:disable Metrics/ClassLength
      # @return [Hash<String, UserType>]
      attr_reader :types

      protected

      # @return [FFI::Polar]
      attr_reader :ffi_polar
      # @return [Hash<Integer, Object>]
      attr_reader :instances
      # @return [Boolean]
      attr_reader :accept_expression

      public

      attr_writer :accept_expression

      def initialize(ffi_polar)
        @ffi_polar = ffi_polar
        @types = {}
        @instances = {}
        @accept_expression = false
      end

      def initialize_copy(other)
        @ffi_polar = other.ffi_polar
        @types = other.types.dup
        @instances = other.instances.dup
      end

      # Fetch a Ruby class from the {#types} cache.
      #
      # @param name [String]
      # @return [Class]
      # @raise [UnregisteredClassError] if the class has not been registered.
      def get_class(name)
        raise UnregisteredClassError, name unless types.key? name

        types[name].klass.get
      end

      # Store a Ruby class in the {#types} cache.
      #
      # @param cls [Class] the class to cache.
      # @param name [String] the name to cache the class as.
      # @return [String] the name the class is cached as.
      # @raise [DuplicateClassAliasError] if attempting to register a class
      # under a previously-registered name.
      def cache_class(cls, name:, fields:, build_query:, combine_query:, exec_query:) # rubocop:disable Metrics/ParameterLists, Metrics/MethodLength
        raise DuplicateClassAliasError.new name: name, old: get_class(name), new: cls if types.key? name

        types[name] = types[cls] = UserType.new(
          name: name,
          klass: PolarClass.new(cls),
          id: cache_instance(cls),
          fields: fields || {},
          combine_query: combine_query,
          exec_query: exec_query,
          build_query: build_query
        )
        name
      end

      def register_mros # rubocop:disable Metrics/AbcSize
        types.values.uniq.each do |typ|
          mro = []
          typ.klass.get.ancestors.each do |a|
            mro.push(types[a].id) if types.key?(a)
          end
          ffi_polar.register_mro(typ.name, mro)
        end
      end

      # Check if an instance exists in the {#instances} cache.
      #
      # @param id [Integer]
      # @return [Boolean]
      def instance?(id)
        case id
        when Integer
          instances.key? id
        else
          instances.value? id
        end
      end

      # Fetch a Ruby instance from the {#instances} cache.
      #
      # @param id [Integer]
      # @return [Object]
      # @raise [UnregisteredInstanceError] if the ID has not been registered.
      def get_instance(id)
        raise UnregisteredInstanceError, id unless instance? id

        instance = instances[id]
        return instance.get if instance.is_a? PolarClass

        instance
      end

      # Cache a Ruby instance in the {#instances} cache, fetching a new id if
      # one isn't provided.
      #
      # @param instance [Object]
      # @param id [Integer] the instance ID. Generated via FFI if not provided.
      # @return [Integer] the instance ID.
      def cache_instance(instance, id: nil)
        id = ffi_polar.new_id if id.nil?
        # Save the instance as a PolarClass if it is a non-anonymous class
        instance = PolarClass.new(instance) if instance.is_a?(Class)
        instances[id] = instance
        id
      end

      # Construct and cache a Ruby instance.
      #
      # @param cls_name [String] name of the instance's class.
      # @param args [Array<Object>] positional args to the constructor.
      # @param kwargs [Hash<String, Object>] keyword args to the constructor.
      # @param id [Integer] the instance ID.
      # @raise [PolarRuntimeError] if instance construction fails.
      # @return [Integer] the instance ID.
      def make_instance(cls_name, args:, kwargs:, id:)
        instance = if kwargs.empty? # This check is for Ruby < 2.7.
                     get_class(cls_name).__send__(:new, *args)
                   else
                     get_class(cls_name).__send__(:new, *args, **kwargs)
                   end
        cache_instance(instance, id: id)
      rescue StandardError => e
        raise PolarRuntimeError, "Error constructing instance of #{cls_name}: #{e}"
      end

      OPS = {
        'Lt' => :<,
        'Gt' => :>,
        'Eq' => :==,
        'Geq' => :>=,
        'Leq' => :<=,
        'Neq' => :!=
      }.freeze

      # Compare two values
      #
      # @param op [String] operation to perform.
      # @param args [Array<Object>] left and right args to operation.
      # @raise [PolarRuntimeError] if operation fails or is unsupported.
      # @return [Boolean]
      def operator(operation, args)
        left, right = args
        op = OPS[operation]
        raise PolarRuntimeError, "Unsupported external operation '#{left.class} #{operation} #{right.class}'" if op.nil?

        begin
          left.__send__ op, right
        rescue StandardError
          raise PolarRuntimeError, "External operation '#{left.class} #{operation} #{right.class}' failed."
        end
      end

      # Check if the left class is more specific than the right class
      # with respect to the given instance.
      #
      # @param instance_id [Integer]
      # @param left_tag [String]
      # @param right_tag [String]
      # @return [Boolean]
      def subspecializer?(instance_id, left_tag:, right_tag:)
        mro = get_instance(instance_id).class.ancestors
        left_index = mro.index(get_class(left_tag))
        right_index = mro.index(get_class(right_tag))
        left_index && right_index && left_index < right_index
      end

      def subclass?(left_tag:, right_tag:)
        get_class(left_tag) <= get_class(right_tag)
      end

      # Check if instance is an instance of class.
      #
      # @param instance [Hash<String, Object>]
      # @param class_tag [String]
      # @return [Boolean]
      def isa?(instance, class_tag:)
        instance = to_ruby(instance)
        cls = get_class(class_tag)
        instance.is_a? cls
      end

      def serialize_types # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
        polar_types = {}
        types.values.uniq.each do |typ|
          tag = typ.name
          fields = typ.fields
          field_types = {}
          fields.each do |k, v|
            field_types[k] =
              if v.is_a? ::Oso::Polar::DataFiltering::Relation
                {
                  'Relation' => {
                    'kind' => v.kind,
                    'other_class_tag' => v.other_type,
                    'my_field' => v.my_field,
                    'other_field' => v.other_field
                  }
                }
              else
                { 'Base' => { 'class_tag' => types[v].name } }
              end
          end
          polar_types[tag] = field_types
        end
        polar_types
      end

      # Turn a Ruby value into a Polar term that's ready to be sent across the
      # FFI boundary.
      #
      # @param value [Object]
      # @return [Hash<String, Object>]
      def to_polar(value) # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/MethodLength, Metrics/PerceivedComplexity
        value = case true # rubocop:disable Lint/LiteralAsCondition
                when value.instance_of?(TrueClass) || value.instance_of?(FalseClass)
                  { 'Boolean' => value }
                when value.instance_of?(Integer)
                  { 'Number' => { 'Integer' => value } }
                when value.instance_of?(Float)
                  if value == Float::INFINITY
                    value = 'Infinity'
                  elsif value == -Float::INFINITY
                    value = '-Infinity'
                  elsif value.nan?
                    value = 'NaN'
                  end
                  { 'Number' => { 'Float' => value } }
                when value.instance_of?(String)
                  { 'String' => value }
                when value.instance_of?(Array)
                  { 'List' => value.map { |el| to_polar(el) } }
                when value.instance_of?(Hash)
                  { 'Dictionary' => { 'fields' => value.transform_values { |v| to_polar(v) } } }
                when value.instance_of?(Predicate)
                  { 'Call' => { 'name' => value.name, 'args' => value.args.map { |el| to_polar(el) } } }
                when value.instance_of?(Variable)
                  # This is supported so that we can query for unbound variables
                  { 'Variable' => value.name }
                when value.instance_of?(Expression)
                  { 'Expression' => { 'operator' => value.operator, 'args' => value.args.map { |el| to_polar(el) } } }
                when value.instance_of?(Pattern)
                  dict = to_polar(value.fields)['value']
                  if value.tag.nil?
                    { 'Pattern' => dict }
                  else
                    { 'Pattern' => { 'Instance' => { 'tag' => value.tag, 'fields' => dict['Dictionary'] } } }
                  end
                else
                  instance_id = nil
                  instance_id = types[value].id if value.is_a?(Class) && types.key?(value)
                  { 'ExternalInstance' => { 'instance_id' => cache_instance(value, id: instance_id), 'repr' => nil } }
                end
        { 'value' => value }
      end

      # Turn a Polar term passed across the FFI boundary into a Ruby value.
      #
      # @param data [Hash<String, Object>]
      # @option data [Integer] :id
      # @option data [Integer] :offset Character offset of the term in its source string.
      # @option data [Hash<String, Object>] :value
      # @return [Object]
      # @raise [UnexpectedPolarTypeError] if type cannot be converted to Ruby.
      def to_ruby(data) # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/MethodLength, Metrics/PerceivedComplexity
        tag, value = data['value'].first
        case tag
        when 'String', 'Boolean'
          value
        when 'Number'
          num = value.values.first
          if value.key? 'Float'
            case num
            when 'Infinity'
              return Float::INFINITY
            when '-Infinity'
              return -Float::INFINITY
            when 'NaN'
              return Float::NAN
            else
              unless value['Float'].is_a? Float # rubocop:disable Metrics/BlockNesting
                raise PolarRuntimeError, "Expected a floating point number, got \"#{value['Float']}\""
              end
            end
          end
          num
        when 'List'
          value.map { |el| to_ruby(el) }
        when 'Dictionary'
          value['fields'].transform_values { |v| to_ruby(v) }
        when 'ExternalInstance'
          get_instance(value['instance_id'])
        when 'Call'
          Predicate.new(value['name'], args: value['args'].map { |a| to_ruby(a) })
        when 'Variable'
          Variable.new(value)
        when 'Expression'
          raise UnexpectedPolarTypeError, tag unless accept_expression

          args = value['args'].map { |a| to_ruby(a) }
          Expression.new(value['operator'], args)
        when 'Pattern'
          case value.keys.first
          when 'Instance'
            tag = value.values.first['tag']
            fields = value.values.first['fields']['fields'].transform_values { |v| to_ruby(v) }
            Pattern.new(tag, fields)
          when 'Dictionary'
            fields = value.values.first['fields'].transform_values { |v| to_ruby(v) }
            Pattern.new(nil, fields)
          else
            raise UnexpectedPolarTypeError, "#{value.keys.first} variant of Pattern"
          end
        else
          raise UnexpectedPolarTypeError, tag
        end
      end

      def enrich_message(msg)
        msg.gsub(/\^\{id: ([0-9]+)\}/) do
          get_instance(Regexp.last_match[1].to_i).to_s
        end
      end
    end
  end
end
